proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,283 @@
+
+Some platforms that support a range of wave sizes might overlap with the range
+ specified as supported by the shader in more than one value.
+In these cases, the runtime has a choice of wave sizes that it probably doesn't
+ have the information needed to choose the best wave size.
+To give the runtime the needed information, it is useful to specify the
+ preferred wave size in addition to the full range of acceptable values.
+By adding an optional third `WaveSize` parameter, shader authors can specify the
+ optimal wave size for a given shader in addition to the range of acceptable
+ sizes.
+
+## Detailed Design
+
+### HLSL additions
+
+The existing `WaveSize` attribute gains a second overload:
+
+```HLSL
+[WaveSize(<minWaveSize>, <maxWaveSize>, [prefWaveSize])]
+void main() ...
+```
+
+Where `minWaveSize` is the minimum wave size supported by the shader
+ representing the beginning of the allowed range,
+ `maxWaveSize` is the maximum wave size supported by the shader
+ representing the end of the allowed range,
+ and `prefWaveSize` is the optional preferred wave size representing the size
+ expected to be the most optimal for this shader.
+
+### DXIL Additions
+
+The existing `WaveSize` value is stored as a metadata scalar constant.
+To store the additional values, an additional metadata tag indicating a tuple of
+ three values representing the parameters to `WaveSize` given by the user.
+In the case where no preferred wave size is specified, the value zero will
+ indicate that nothing was specified.
+
+### SPIR-V Additions
+
+To make use of this attribute, SPIRV would have to represent it in new SPIRV
+ operations that drivers can read to determine what the values are an how to
+ use to them to choose the best wave size for a given shader.
+
+### Diagnostic Changes
+
+#### New Errors
+
+These are where new or slightly altered errors are produced:
+
+* If any of the parameters of `WaveSize` are not literal power of two integers
+  between 4 and 128 inclusive.
+* If the minimum wave size value is greater than or equal to the max wave size
+ as respectively specified by the first and second parameters in the `WaveSize`
+ attribute.
+* If the preferred wave size value indicated by the third parameter of
+ `WaveSize` is not in the range specified by the first two parameters.
+* If multiple `WaveSize` attributes are applied to the same entry point,
+ regardless of what overload they are, the existing error indicating an
+ attribute conflict is produced.
+* If more than three or fewer than one parameter is applied to `WaveSize`.
+* If negative values are provided for any of the `WaveSize` parameters.
+* If float values are provided for any of the `WaveSize` parameters.
+* If non-numerical values are provided for any of the `WaveSize` parameters.
+* If non-literal variables are provided for any of the `WaveSize` parameters.
+* If any of the parameters to `WaveSize` are less than 4 or greater than 128.
```

Comment:
Aren't these points already covered by the first one:
```
* If any of the parameters of `WaveSize` are not literal power of two integers
  between 4 and 128 inclusive.
```

Reply:
I'm wondering why 1 and 2 aren't valid wave sizes?  (I suspect I'm missing something obvious)

Reply:
> I'm wondering why 1 and 2 aren't valid wave sizes? (I suspect I'm missing something obvious)

IIRC, we require wave size of 4 so that derivative operations and quad ops work.

Reply:
There's a separate cap for those.  Not allowing WaveSize=1 prohibits a single threaded implementation, that I imagine could be interesting for some debugging scenarios.  Maybe not important, this is just ringing my arbitrary restriction bell.

Reply:
We have a cap bit for the `Quad` ops, but not for the derivative functions (`ddx`, `ddy`, etc).

Reply:
Wait... nvm. We do have a cap bit for derivatives in mesh and amplification shaders. Please ignore me.

Reply:
I think @tex3d's feedback here still stands though that this is a duplicate test criteria based on the first bullet.

Reply:
Ah yes. I'd missed that in the interchange. Done.

Reply:
It looks like most of the lines I highlighted are still there.  Did you miss these, or forget to push an update?

Reply:
I missed some. Fixed.